Source/WebCore/dom/WindowEventLoop.h

```diff
@@ -53,6 +53,12 @@ class WindowEventLoop final : public EventLoop {
 
     CustomElementQueue& backupElementQueue();
 
+    void didScheduleRenderingUpdate() { m_hasARenderingOpportunity = true; }
+    void didFinishRenderingUpdate() { m_hasARenderingOpportunity = false; }
```

Comment:
So we do not run idle callback when next rendering is already scheduled (and waiting), makes sense.
Is it better to ask to RenderingUpdateScheduler directly? (For example, taking `Page&` in `opportunisticallyRunIdleCallbacks`, threading it to `shouldEndIdlePeriod` and ask to `RenderingUpdateScheduler` instead of configuring this bool in each WindowEventLoop.

Reply:
So one subtle thing here is that we don't use Page's `RenderingUpdateScheduler` to schedule rendering update in WebKit2. Instead, ChromeClient in WebKit2 schedules rendering update and we early exit in `scheduleRenderingUpdateInternal`.